File: backend/app/api/races.py
"""Race API routes."""

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas import (
    RaceCreate,
    RaceListResponse,
    RaceResponse,
    RaceUpdate,
)
from app.services import RaceService

router = APIRouter(prefix="/races", tags=["races"])


@router.get("", response_model=RaceListResponse)
async def get_races(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    grade: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Get all races with pagination."""
    service = RaceService(db)
    races, total = await service.get_races(skip=skip, limit=limit, grade=grade)
    return RaceListResponse(items=races, total=total)


@router.get("/upcoming", response_model=list[RaceResponse])
async def get_upcoming_races(
    limit: int = Query(10, ge=1, le=50),
    db: AsyncSession = Depends(get_db),
):
    """Get upcoming races."""
    service = RaceService(db)
    return await service.get_upcoming_races(limit)


@router.get("/search", response_model=list[RaceResponse])
async def search_races(
    name: str = Query(..., min_length=1),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """Search races by name."""
    service = RaceService(db)
    return await service.search_races(name, limit)


@router.get("/by-date", response_model=list[RaceResponse])
async def get_races_by_date_range(
    start_date: date = Query(...),
    end_date: date = Query(...),
    grade: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    """Get races within a date range."""
    if start_date > end_date:
        raise HTTPException(status_code=400, detail="start_date must be before end_date")

    service = RaceService(db)
    return await service.get_races_by_date_range(start_date, end_date, grade)


@router.get("/{race_id}", response_model=RaceResponse)
async def get_race(
    race_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get a race by ID."""
    service = RaceService(db)
    race = await service.get_race(race_id)
    if not race:
        raise HTTPException(status_code=404, detail="Race not found")
    return race


@router.post("", response_model=RaceResponse, status_code=201)
async def create_race(
    data: RaceCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create a new race."""
    service = RaceService(db)
    return await service.create_race(data)


@router.put("/{race_id}", response_model=RaceResponse)
async def update_race(
    race_id: int,
    data: RaceUpdate,
    db: AsyncSession = Depends(get_db),
):
    """Update a race."""
    service = RaceService(db)
    race = await service.update_race(race_id, data)
    if not race:
        raise HTTPException(status_code=404, detail="Race not found")
    return race


@router.delete("/{race_id}", status_code=204)
async def delete_race(
    race_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Delete a race."""
    service = RaceService(db)
    deleted = await service.delete_race(race_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Race not found")

File: frontend/src/api/races.ts
// Race API functions

import apiClient from './client';
import type { Race, RaceCreate, RaceListResponse, RaceUpdate } from '../types';

export interface RegisterRaceRequest {
  netkeiba_race_id: string;
  fetch_odds?: boolean;
  fetch_running_styles?: boolean;
}

export interface RegisterRaceResponse {
  success: boolean;
  message: string;
  race_id: number | null;
  race_name: string | null;
  entries_count: number;
  odds_updated: number;
}

export const racesApi = {
  // Get all races with pagination
  getAll: async (params?: {
    skip?: number;
    limit?: number;
    grade?: string;
  }): Promise<RaceListResponse> => {
    const response = await apiClient.get('/races', { params });
    return response.data;
  },

  // Get upcoming races
  getUpcoming: async (limit = 10): Promise<Race[]> => {
    const response = await apiClient.get('/races/upcoming', {
      params: { limit },
    });
    return response.data;
  },

  // Search races by name
  search: async (name: string, limit = 20): Promise<Race[]> => {
    const response = await apiClient.get('/races/search', {
      params: { name, limit },
    });
    return response.data;
  },

  // Get races by date range
  getByDateRange: async (
    startDate: string,
    endDate: string,
    grade?: string
  ): Promise<Race[]> => {
    const response = await apiClient.get('/races/by-date', {
      params: {
        start_date: startDate,
        end_date: endDate,
        grade,
      },
    });
    return response.data;
  },

  // Get single race by ID
  getById: async (raceId: number): Promise<Race> => {
    const response = await apiClient.get(`/races/${raceId}`);
    return response.data;
  },

  // Create new race
  create: async (data: RaceCreate): Promise<Race> => {
    const response = await apiClient.post('/races', data);
    return response.data;
  },

  // Update race
  update: async (raceId: number, data: RaceUpdate): Promise<Race> => {
    const response = await apiClient.put(`/races/${raceId}`, data);
    return response.data;
  },

  // Delete race
  delete: async (raceId: number): Promise<void> => {
    await apiClient.delete(`/races/${raceId}`);
  },

  // Register race from netkeiba ID
  registerFromNetkeiba: async (data: RegisterRaceRequest): Promise<RegisterRaceResponse> => {
    const response = await apiClient.post('/fetch/register', data);
    return response.data;
  },
};
